src/main/cpp/src/hive_hash.cu

```diff
@@ -0,0 +1,237 @@
+  auto* p_int     = reinterpret_cast<int32_t const*>(&normalized);
+  return compute_int(*p_int);
+}
+
+template <>
+hive_hash_value_t __device__ inline hive_hash_function<double>::operator()(double const& key) const
+{
+  auto normalized = spark_rapids_jni::normalize_nans(key);
+  auto* p_long    = reinterpret_cast<int64_t const*>(&normalized);
+  return compute_long(*p_long);
+}
+
+template <>
+hive_hash_value_t __device__ inline hive_hash_function<cudf::timestamp_D>::operator()(
+  cudf::timestamp_D const& key) const
+{
+  auto* p_int = reinterpret_cast<int32_t const*>(&key);
+  return compute_int(*p_int);
+}
+
+/**
+ * @brief Computes the hash value of a row in the given table.
+ *
+ * This functor produces the same result as "HiveHash" in Spark for supported types.
+ *
+ * @tparam hash_function Hash functor to use for hashing elements. Must be hive_hash_function.
+ * @tparam Nullate A cudf::nullate type describing whether to check for nulls.
+ */
+template <template <typename> class hash_function, typename Nullate>
+class hive_device_row_hasher {
+ public:
+  CUDF_HOST_DEVICE hive_device_row_hasher(Nullate check_nulls, cudf::table_device_view t) noexcept
+    : _check_nulls{check_nulls}, _table{t}
+  {
+    // Error out if passed an unsupported hash_function
+    static_assert(std::is_base_of_v<hive_hash_function<int>, hash_function<int>>,
+                  "hive_device_row_hasher only supports the 'hive_hash_function' hash function");
+  }
+
+  /**
+   * @brief Return the hash value of a row in the given table.
+   *
+   * @param row_index The row index to compute the hash value of
+   * @return The hash value of the row
+   */
+  __device__ auto operator()(cudf::size_type row_index) const noexcept
+  {
+    return cudf::detail::accumulate(
+      _table.begin(),
+      _table.end(),
+      HIVE_INIT_HASH,
+      cuda::proclaim_return_type<hive_hash_value_t>(
+        [row_index, nulls = this->_check_nulls] __device__(auto hash, auto column) {
+          auto cur_hash = cudf::type_dispatcher(
+            column.type(), element_hasher_adapter<hash_function>{nulls}, column, row_index);
+          return HIVE_HASH_FACTOR * hash + cur_hash;
+        }));
+  }
+
+ private:
+  /**
+   * @brief Computes the hash value of an element in the given column.
+   *
+   * Only supported non nested types now
+   */
+  template <template <typename> class hash_fn>
+  class element_hasher_adapter {
+   public:
+    __device__ element_hasher_adapter(Nullate check_nulls) noexcept : _check_nulls(check_nulls) {}
+
+    using hash_functor = cudf::experimental::row::hash::element_hasher<hash_fn, Nullate>;
+
+    template <typename T, CUDF_ENABLE_IF(not cudf::is_nested<T>())>
+    __device__ hive_hash_value_t operator()(cudf::column_device_view const& col,
+                                            cudf::size_type row_index) const noexcept
+    {
+      auto const hasher = hash_functor{_check_nulls, HIVE_INIT_HASH, HIVE_INIT_HASH};
```

Comment:
`hasher` seems to be constant thus it can be a member variable and initialized upon class construction.

Reply:
Doing so, we can remove `_check_nulls` from this class.

Reply:
good idea, done